File: 0677-map-sum-pairs/0677-map-sum-pairs.cpp
struct TrieNode {
    TrieNode* child[26] = {};
    int sum = 0; // Store the sum of values of all strings go through this node.
};

class MapSum { // 0 ms, faster than 100.00%
public:
    unordered_map<string, int> map;
    TrieNode trieRoot;
    
    void insert(const string& key, int val) {
        int diff = val - map[key];
        TrieNode* curr = &trieRoot;
        for (char c : key) {
            c -= 'a';
            if (curr->child[c] == nullptr) curr->child[c] = new TrieNode();
            curr = curr->child[c];
            curr->sum += diff;
        }
        map[key] = val;
    }
    int sum(const string& prefix) {
        TrieNode* curr = &trieRoot;
        for (char c : prefix) {
            c -= 'a';
            if (curr->child[c] == nullptr) return 0;
            curr = curr->child[c];
        }
        return curr->sum;
    }
};

File: 0677-map-sum-pairs/0677-map-sum-pairs.java
class TrieNode {
    TrieNode[] child = new TrieNode[26];
    int sum = 0; 
}

class MapSum { 
    HashMap<String, Integer> map = new HashMap<>();
    TrieNode trieRoot = new TrieNode();

    public void insert(String key, int val) {
        int diff = val - map.getOrDefault(key, 0);
        TrieNode curr = trieRoot;
        for (char c : key.toCharArray()) {
            c -= 'a';
            if (curr.child[c] == null) curr.child[c] = new TrieNode();
            curr = curr.child[c];
            curr.sum += diff;
        }
        map.put(key, val);
    }
    public int sum(String prefix) {
        TrieNode curr = trieRoot;
        for (char c : prefix.toCharArray()) {
            c -= 'a';
            if (curr.child[c] == null) return 0;
            curr = curr.child[c];
        }
        return curr.sum;
    }
}